gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/proc/DagProcUtils.java

```diff
@@ -0,0 +1,118 @@
+      // Ensure that we do not double increment for flows that are retried
+      if (DagManagerUtils.getJobExecutionPlan(dagNode).getCurrentAttempts() == 1) {
+        dagManagementStateStore.getDagManagerMetrics().incrementRunningJobMetrics(dagNode);
+      }
+      // Submit the job to the SpecProducer, which in turn performs the actual job submission to the SpecExecutor instance.
+      // The SpecProducer implementations submit the job to the underlying executor and return when the submission is complete,
+      // either successfully or unsuccessfully. To catch any exceptions in the job submission, the DagManagerThread
+      // blocks (by calling Future#get()) until the submission is completed.
+      dagManagementStateStore.tryAcquireQuota(Collections.singleton(dagNode));
+      Future<?> addSpecFuture = producer.addSpec(jobSpec);
+      // todo - we should add future.get() instead of the complete future into the JobExecutionPlan
+      dagNode.getValue().setJobFuture(com.google.common.base.Optional.of(addSpecFuture));
+      addSpecFuture.get();
+      jobExecutionPlan.setExecutionStatus(ExecutionStatus.ORCHESTRATED);
+      jobMetadata.put(TimingEvent.METADATA_MESSAGE, producer.getExecutionLink(addSpecFuture, specExecutorUri));
+      // Add serialized job properties as part of the orchestrated job event metadata
+      jobMetadata.put(JobExecutionPlan.JOB_PROPS_KEY, dagNode.getValue().toString());
+      jobOrchestrationTimer.stop(jobMetadata);
+      log.info("Orchestrated job: {} on Executor: {}", DagManagerUtils.getFullyQualifiedJobName(dagNode), specExecutorUri);
+      dagManagementStateStore.getDagManagerMetrics().incrementJobsSentToExecutor(dagNode);
+      dagManagementStateStore.addDagNodeState(dagNode, dagId);
+    } catch (Exception e) {
+      TimingEvent jobFailedTimer = DagProc.eventSubmitter.getTimingEvent(TimingEvent.LauncherTimings.JOB_FAILED);
+      String message = "Cannot submit job " + DagManagerUtils.getFullyQualifiedJobName(dagNode) + " on executor " + specExecutorUri;
+      log.error(message, e);
+      jobMetadata.put(TimingEvent.METADATA_MESSAGE, message + " due to " + e.getMessage());
+      if (jobFailedTimer != null) {
+        jobFailedTimer.stop(jobMetadata);
+      }
+      try {
+        // when there is no exception, quota will be released in job status monitor or re-evaluate dag proc
+        dagManagementStateStore.releaseQuota(dagNode);
+      } catch (IOException ex) {
+        log.error("Could not release quota while handling e", ex);
+      }
+      throw new RuntimeException(e);
+    }
+  }
+
+  public static boolean hasRunningJobs(DagManager.DagId dagId, DagManagementStateStore dagManagementStateStore)
+      throws IOException {
+    List<Dag.DagNode<JobExecutionPlan>> dagNodes = dagManagementStateStore.getDagNodes(dagId);
+    return dagNodes != null && !dagNodes.isEmpty();
+  }
```

Comment:
since the impl merely invokes a DMSS method and then analyzes the result, I wonder: would this better fit as a method on DMSS?